clients/include/norm.hpp

```diff
@@ -21,28 +28,260 @@
+
+    double error = xlang(&norm_type, &M, &N, hGPU, &lda, work) / cpu_norm;
+
+    return error;
+}
 
-/*! \brief  Template: norm check for hermitian/symmetric Matrix: half/float/double/complex */
+template <>
+inline double norm_check_general<rocblas_half, nullptr>(char          norm_type,
+                                                        rocblas_int   M,
+                                                        rocblas_int   N,
+                                                        rocblas_int   lda,
+                                                        rocblas_half* hCPU,
+                                                        rocblas_half* hGPU)
+{
+    // norm type can be 'O', 'I', 'F', 'o', 'i', 'f' for one, infinity or Frobenius norm
+    // one norm is max column sum
+    // infinity norm is max row sum
+    // Frobenius is l2 norm of matrix entries
+    host_vector<double> hCPU_double(N * lda);
+    host_vector<double> hGPU_double(N * lda);
+
+    for(rocblas_int i = 0; i < N * lda; i++)
+    {
+        hCPU_double[i] = double(half_to_float(hCPU[i]));
+        hGPU_double[i] = double(half_to_float(hGPU[i]));
+    }
+
+    return norm_check_general(norm_type, M, N, lda, hCPU_double.data(), hGPU_double.data());
+}
 
 template <typename T>
-double norm_check_symmetric(
-    char norm_type, char uplo, rocblas_int N, rocblas_int lda, T* hCPU, T* hGPU);
+inline double norm_check_general(char        norm_type,
+                                 rocblas_int M,
+                                 rocblas_int N,
+                                 rocblas_int lda,
+                                 rocblas_int stride_a,
+                                 rocblas_int batch_count,
+                                 T*          hCPU,
+                                 T*          hGPU)
+{
+    // norm type can be O', 'I', 'F', 'o', 'i', 'f' for one, infinity or Frobenius norm
+    // one norm is max column sum
+    // infinity norm is max row sum
+    // Frobenius is l2 norm of matrix entries
+    //
+    // use triangle inequality ||a+b|| <= ||a|| + ||b|| to calculate upper limit for Frobenius norm
+    // of strided batched matrix
+
+    double cumulative_error = 0.0;
+
+    for(rocblas_int i = 0; i < batch_count; i++)
+    {
+        auto index = i * stride_a;
+
+        auto error = norm_check_general(norm_type, M, N, lda, hCPU + index, hGPU + index);
+
+        if(norm_type == 'F' || norm_type == 'f')
+        {
+            cumulative_error += error;
+        }
+        else if(norm_type == 'O' || norm_type == 'o' || norm_type == 'I' || norm_type == 'i')
+        {
+            cumulative_error = cumulative_error > error ? cumulative_error : error;
+        }
+    }
+
+    return cumulative_error;
+}
+
+/* ============== Norm Check for Symmetric Matrix ============= */
+/*! \brief compare the norm error of two hermitian/symmetric matrices hCPU & hGPU */
+
+template <typename T, typename std::enable_if<!is_complex<T>>::type* = nullptr>
```

Comment:
Pointers to `void` are not allowed as template arguments. It should be changed to:
`template <typename T, typename std::enable_if<!is_complex<T>, int>::type = 0>`

Reply:
understand. update it in commit "update complex change 2".